File: LC/Medium/207.py
class Solution:
    def canFinish(self, n: int, prerequisites: List[List[int]]) -> bool:
        adj_list = defaultdict(list)
        in_degree = [0] * n
        
        for parent, child in prerequisites:
            adj_list[parent].append(child)
            in_degree[child] += 1
            
        sources = [i for i in range(n) if in_degree[i] == 0]
        
        courses = 0
        while sources:
            curr = sources.pop(0)
            courses += 1
            
            for neighbour in adj_list[curr]:
                in_degree[neighbour] -= 1
                if in_degree[neighbour] == 0:
                    sources.append(neighbour)
        print(courses)
        return courses == n

File: LC/C++/Medium/207.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        std::unordered_map<int, vector<int>> adjList{};
        std::vector<int> inNodes(numCourses);

        for (vector<int>& req : prerequisites) {
            int a = req[0];
            int b = req[1];
            adjList[b].push_back(a);
            inNodes[a]++;
        }

        std::queue<int> sources;
        for (int i = 0; i < numCourses; ++i) {
            if (inNodes[i] == 0) {
                sources.push(i);
            }
        }

        while (!sources.empty()) {
            int curr = sources.front();
            sources.pop();

            for (int course : adjList[curr]) {
                inNodes[course]--;
                if (inNodes[course] == 0) {
                    sources.push(course);
                } 
            }
            numCourses--;
        }

        return numCourses == 0;
    }
};